### PR TITLE
Nmi-ari and sim modules

### PR DESCRIPTION
**Pull request by atalon-lip6**
This PR adds two modules:
NMI-ARI to compare two clusterings
SIM to compare two distance matrices (for example in low and high dimensions).

### CODE REVIEW

**Comment by julien-tierny** [comment]
hi @atalon-lip6 ,
thanks for the PR. can you please fix the fail checks before I proceed to the review?
thanks!

**Comment by atalon-lip6** [comment]
> hi @atalon-lip6 , thanks for the PR. can you please fix the fail checks before I proceed to the review? thanks!

Hello! This should be okay now.

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    sumLin[i1] = sum;
+  }
+
+  return 0;
+}
+
+int ttk::ClusteringMetrics::computeARI(
+  std::vector<std::vector<int>> &contingencyMatrix,
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: I believe this parameter should be `const` (it seems to be used in read-only mode).

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                     - (sumNChoose2_1 * sumNChoose2_2) / nChoose2(nPoint);
+  double denominator = 0.5 * (sumNChoose2_1 + sumNChoose2_2)
+                       - (sumNChoose2_1 * sumNChoose2_2) / nChoose2(nPoint);
+  ariValue = numerator / denominator;
+
+  return 0;
+}
+
+int ttk::ClusteringMetrics::computeNMI(
+  std::vector<std::vector<int>> &contingencyMatrix,
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: I believe this parameter should be `const` (it seems to be used in read-only mode).

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  return 0;
+}
+
+int ttk::ClusteringMetrics::computeARI(
+  std::vector<std::vector<int>> &contingencyMatrix,
+  const std::vector<int> &sumLin,
+  const std::vector<int> &sumCol,
+  int nPoint,
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: this parameter may be set to `const` (to explicitly mark it as read-only).

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  ariValue = numerator / denominator;
+
+  return 0;
+}
+
+int ttk::ClusteringMetrics::computeNMI(
+  std::vector<std::vector<int>> &contingencyMatrix,
+  const std::vector<int> &sumLin,
+  const std::vector<int> &sumCol,
+  int nPoint,
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: this parameter may be set to `const` (to explicitly mark it as read-only).

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+int ttk::ClusteringMetrics::execute(const std::vector<int> &clustering1,
+                                    const std::vector<int> &clustering2,
+                                    double &nmiValue,
+                                    double &ariValue) const {
+  ttk::Timer timer;
+  size_t n = clustering1.size();
+
+#ifndef TTK_ENABLE_KAMIKAZE
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: the flag `TTK_ENABLE_KAMIKAZE` is meant to only be set around consistency checks in low-level portions of the code (typically in the directory `core/base`).

in general, `printMsg` calls should be left out of `TTK_ENABLE_KAMIKAZE` code blocks.
then, the message will only be displayed if the object is called with the appropriate debug level.

in this example, I'd suggest to remove the lines 217 to 225 (I guess they're leftovers from the helloworld example)

File: core/base/clusteringMetrics/ClusteringMetrics.cpp [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+  std::vector<std::vector<int>> contingencyMatrix;
+  std::vector<int> sumLines, sumColumns;
+  computeContingencyTables(
+    clustering1, clustering2, contingencyMatrix, sumLines, sumColumns);
+
+  computeARI(contingencyMatrix, sumLines, sumColumns, n, ariValue);
+  computeNMI(contingencyMatrix, sumLines, sumColumns, n, nmiValue);
+
+#ifndef TTK_ENABLE_KAMIKAZE
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.cpp]: I would remove the `TTK_ENABLE_KAMIKAZE` macro here.
the messages below may be important for the user.

File: core/base/clusteringMetrics/ClusteringMetrics.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+/// \date January 2022
+///
+/// This module defines the %ClusteringMetrics class that computes both the NMI
+/// (normalized mutual information) and the ARI (adjusted rand index) scores
+/// between two clustering of the same points.
+///
+/// \b Related \b publication: \n
+/// 'Principal Geodesic Analysis of Merge Trees (and Persistence Diagrams)'
+/// Mathieu Pont, Jules Vidal and Julien Tierny.
+/// TTK Publications.
```
Comment (julien-tierny) [inline core/base/clusteringMetrics/ClusteringMetrics.h]: IEEE Transactions on Visualization and Computer Graphics, 2023.

File: core/base/distanceMatrixDistorsion/DistanceMatrixDistorsion.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+int ttk::DistanceMatrixDistorsion::execute(
+  const std::vector<std::vector<double>> &highDistMatrix,
+  const std::vector<std::vector<double>> &lowDistMatrix,
+  double &distorsionValue,
+  std::vector<double> &distorsionVerticesValues) const {
+  ttk::Timer timer;
+  auto n = highDistMatrix.size();
+
+#ifndef TTK_ENABLE_KAMIKAZE
```
Comment (julien-tierny) [inline core/base/distanceMatrixDistorsion/DistanceMatrixDistorsion.cpp]: same remark as before regarding `TTK_ENABLE_KAMIKAZE`.
I would remove the block line 21 to 29.

File: core/base/distanceMatrixDistorsion/DistanceMatrixDistorsion.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+      sum += diff2;
+    }
+    double sumHarmonized = sum / maxi;
+    distorsionVerticesValues[i] = 1 - sumHarmonized / n;
+    totalSum += 1 - sumHarmonized / n;
+  }
+
+  distorsionValue = totalSum / n;
+
+#ifndef TTK_ENABLE_KAMIKAZE
```
Comment (julien-tierny) [inline core/base/distanceMatrixDistorsion/DistanceMatrixDistorsion.cpp]: same remark as before about timings (no `TTK_ENABLE_KAMIKAZE` safeguard here).

File: core/vtk/ttkDistanceMatrixDistorsion/ttkDistanceMatrixDistorsion.cpp [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+  // Selecting the columns names if a regexp was given.
+  if(SelectFieldsWithRegexpHigh)
+    fillWithInputColumns(inputHigh, RegexpStringHigh, ScalarFieldsHigh);
+  if(SelectFieldsWithRegexpLow)
+    fillWithInputColumns(inputLow, RegexpStringLow, ScalarFieldsLow);
+
+  const size_t nRowsHigh = inputHigh->GetNumberOfRows();
+
+#ifndef TTK_ENABLE_KAMIKAZE
```
Comment (julien-tierny) [inline core/vtk/ttkDistanceMatrixDistorsion/ttkDistanceMatrixDistorsion.cpp]: I would remove the safeguard `TTK_ENABLE_KAMIKAZE` has this part of the code is not low-level (it is not continuously called by a calling program) and informing the user about a potentially invalid input is useful.

**Review by julien-tierny — CHANGES_REQUESTED** [review]
ok, that looks great and very  polished. thanks!

thanks for considering my few minor remarks.

**Comment by atalon-lip6** [comment]
I implemented the modifications, this should be good now @julien-tierny

**Comment by julien-tierny** [comment]
alright, looks good. thanks!

**Pull request merged**